packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1466,11 +1478,126 @@ export class ContainerRuntime extends EventEmitter
         const summary = builder.getSummaryTree();
         return { ...summary, id: "" };
     }
 
+    public runGarbageCollection(): string[] {
+        // Maintains a map of dataStoreId to an IReferences object. The IReferences object keeps track of whether
+        // the data store has been visited (to detect loops) and the fluid objects in it that are referenced.
+        const dataStoreReferencesMap: Map<string, IReferences> = new Map();
+
+        // A stack of routes that are referenced through handles in DDSes.
+        const referencedRoutesStack: string[] = [];
+
+        // Start with the root data stores. Add all the fluid objects referenced by the root data stores in to the
+        // referencedObjectsStack.
+        this.contexts.forEach((value: FluidDataStoreContext) => {
+            if (value.isRootStore) {
+                // Ignore the scheduler.
+                if (value.id === taskSchedulerId) {
+                    return;
+                }
+
+                // Get the fluid objects referenced by this data store and add them to the referencedObjectsStack.
+                const referencedRoutes = value.getReferencedRoutes();
+                referencedRoutesStack.push(...referencedRoutes);
+
+                // Set the IReferences object in the dataStoreReferencesMap for this data store. Mark it as visited.
+                dataStoreReferencesMap.set(value.id, { visited: true, referencedRoutes: new Set<string>() });
+            }
+        });
+
+        // Do a Bread First Search on the referencedRoutesStack.
+        while (referencedRoutesStack.length > 0) {
+            // Get the next route.
+            const nextRoute = referencedRoutesStack.shift();
+            assert(nextRoute);
+
+            // Get the dataStoreId from the referenced object.
+            const dataStoreId = nextRoute.split("/")[1];
+            // If there is no reference object for it in the dataStoreReferencesMap, it means this is the first time we
+            // are seeing it. Creata a reference object for it.
+            let references = dataStoreReferencesMap.get(dataStoreId);
+            if (references === undefined) {
+                references = {
+                    visited: false,
+                    referencedRoutes: new Set<string>(),
+                };
+            }
+
+            const visited: boolean = references.visited;
+            // Keeps track of whether the entry for this data store in the dataStoreReferencesMap changes.
+            let dataStoreEntryChanged: boolean = false;
+
+            // Get the path of the referenced object relative to the dataStoreId and add it to the
+            // referencedFluidObjects in its references.
+            const dataStoreIdRelativePath = nextRoute.slice(dataStoreId.length + 1);
+            if (dataStoreIdRelativePath !== "") {
+                assert(references.referencedRoutes);
+                references.referencedRoutes.add(dataStoreIdRelativePath);
```

Comment:
Something to keep in mind: With current approach, we will never detect cycles involving DDSs. I.e. if data stores are references, but their DDSs create unreferenced (from outside) cycle, then they will be considered referenced and never collected.
So any future potential to collect DDSs would require substantially different algorithm. Maybe we from the start say that DDS collection is out of the window.
That all said, giving info to data store on unused (or used) paths is still valuable, as (IMHO) we have to figure out a way for data stored to collect sub-DDS objects, like at-mention where all at-mentions are part of same DDS. This requires file change (i.e. ops on the wire), but it's something that we will need to figure out in the future.